transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
-			page = page.nextPage
+		version := make([]byte, versionLength)
+		_, err4 := buffer.Read(version)
+
+		// Unmarshal instructions
+		var instructionsLength uint64
+		err5 := binary.Read(buffer, binary.LittleEndian, &instructionsLength)
+
+		instructions := make([]byte, instructionsLength)
+		_, err6 := buffer.Read(instructions)
+
+		// Check if any errors occured
+		if err1 != nil || err2 != nil || err3 != nil || err4 != nil || err5 != nil || err6 != nil {
+			return nil, errors.Compose(errors.New("Failed to unmarshal updates"), err1, err2, err3, err4, err5, err6)
 		}
 
-		// Commit transaction
-		notifyChannel <- t.commit()
-	}()
-	return notifyChannel
+		update.Name = string(name)
+		update.Version = string(version)
+		update.Instructions = instructions
+
+		updates = append(updates, update)
+	}
+
+	return updates, nil
 }
 
-// MarshalJSON marshales a transaction.
-func (t Transaction) checksum() ([crypto.HashSize]byte, error) {
+// threadedInitTransaction reserves pages of the wal, marshalls the
+// transactions's updates into a payload and splits the payload equally among
+// the pages. Once finished those pages are written to disk and the transaction
+// is committed.
+func threadedInitTransaction(t *Transaction) {
+	defer close(t.initComplete)
+
+	// Marshal all the updates to get their total length on disk
+	data, err := marshalUpdates(t.Updates)
+	if err != nil {
+		t.initComplete <- build.ExtendErr("could not marshal update", err)
+		return
+	}
+
+	// Find out how many pages are needed for the update
+	requiredPages := uint64(float64(len(data))/float64(maxPayloadSize+8) + 1)
+
+	// Get the pages from the wal
+	reservedPages := t.wal.reservePages(requiredPages)
+
+	// Set the fields of each page
+	pages := make([]page, requiredPages)
+	for i := uint64(0); i < requiredPages; i++ {
+		// Set offset according to the index in reservedPages
+		pages[i].offset = reservedPages[i]
+
+		// Set nextPage if the current page isn't the last one
+		// otherwise let it be nil
+		if i+1 < requiredPages {
+			pages[i].nextPage = &pages[i+1]
+		}
+
+		// Set pageStatus of the first page to pageStatusWritten
+		if i == 0 {
+			pages[i].pageStatus = pageStatusWritten
+		} else {
+			pages[i].pageStatus = pageStatusOther
+		}
+
+		// Copy part of the update into the payload
+		payloadsize := maxPayloadSize
+		if len(data[i*maxPayloadSize:]) < payloadsize {
+			payloadsize = len(data[i*maxPayloadSize:])
+		}
+		pages[i].payload = make([]byte, payloadsize)
+		copy(pages[i].payload, data[i*maxPayloadSize:])
+	}
+
+	// Set the first and final page of the transaction
+	t.firstPage = &pages[0]
+	t.finalPage = &pages[len(pages)-1]
+
+	// write the pages to disk and set the pageStatus
 	page := t.firstPage
-	var data []byte
 	for page != nil {
-		// make sure that the checksum field is just zeros before calculating the checksum
-		copy(page.transactionChecksum[:], make([]byte, crypto.HashSize))
-
-		bytes, err := page.Marshal()
+		err := page.Write(t.wal.logFile)
```

Comment:
This is an incorrect implementation of the `Write` function. Since it's not a formal copy of the io.Writer interface, it should probably be renamed `WriteToFile` to avoid confusion.

I am thinking that we should adjust the marshalling and unmarshalling and writing functions to all operate on just a first page, and then to automatically include the remaining pages as needed.

Reply:
So you'd just call `page.WriteToFile(t.wal.logFile)` one time, and then inside the `WriteToFile` function, it would write all of the secondary pages.